### ISSUE TITLE
Promise.prototype.done

### ISSUE THREAD

**Issue by medikoo** [domenic/promises-unwrapping#19]
With current design, the only way to process resolved value and be sure that unhandled exceptions are exposed, is via following

``` javascript
promise.then(function (value) {
  setTimeout(function () {
    // Process value
   }, 0);
}, function (err) {
  setTimeout(function () {
    // Handle error
   }, 0);
});
```

Which is poor solution for many reasons:
- Affects performance:
  - Creates promise objects we don't need
  - forces processing to further tick
- Unintuitive
- Verbose

In many promise implementations, there's already `promise.done`, which doesn't create any new promise, and invokes callbacks naturally (outside of try/catch clause).

As `done` cannot be implemented on top of available API, I think it's very important to include it in spec.

_As side note, why do `Promise.prototype.catch` is specified? It can be very easily configured on top of the rest of the API, and it's usability is low._

**Comment by annevk** [comment]
The outcome of the previous TC39 discussion was to get rid of done() for now. It might resurface one way or another, but adding it back before revisiting that discussion seems bad, and we can easily revisit that after we ship the basic design.

**Comment by medikoo** [comment]
@annevk Thanks for clarification. I just wanted to point that without that, spec (even in most basic form) shouldn't be perceived as complete.

**Comment by domenic** [comment]
`done`'s job will be done by integrating unhandled rejection tracking functionality into dev tools. Most TC39ers, from what I understand, as well as myself, perceive that as enough for the spec to be complete.

Furthermore, there is no need for implementations to create promise objects only to have them immediately GCed. Unlike user-space promise implementations, that is easily detectable; such `then` invocations whose return values are ignored can simply not create a new promise object.

**Comment by medikoo** [comment]
> done's job will be done by integrating unhandled rejection tracking functionality into dev tools. Most TC39ers, from what I understand, as well as myself, perceive that as enough for the spec to be complete.

_Monitoring_ solution significantly raises complexity of implementation, and brings many other issues. I haven't seen it yet either implemented or specified in a way that would really solve above issue.

It's really hard to understand for me, why you advocate after something that is difficult to implement, not logical and not natural for JavaScript, when real solution can be so simple, and as far as I remember from [discussion](http://mozilla.6506.n7.nabble.com/Where-d-Promise-done-go-tt281461.html) on _es-discuss_, many were concerned about that.

> Furthermore, there is no need for implementations to create promise objects only to have them immediately GCed. Unlike user-space promise implementations, that is easily detectable

Can you explain exactly, how? Static code analysis?

**Comment by annevk** [comment]
Implementations can report unhandled exceptions during GC to the error console.

**Comment by medikoo** [comment]
@annevk You mean that when promise would be GC'ed, implementation would check whether reject value was retrieved and if not, it would report to reject value to error console (?)

What if implementation would completely implement this spec, but won't provide this functionality?
I hope you understand that such implementation would turn out as hardly usable.

...and again it's not error handling that JS programmers are used to, it can be simple and should be in my opinion.

**Comment by annevk** [comment]
If implementations do not provide adequate debugging, they'll be used less. This is the same for any number of features on the platform. And simple is not necessarily good here, as people will use .then() and forget about things.

**Comment by medikoo** [comment]
@annevk only with `done` programmer can have full control over error handling, and promise libraries should provide that. 

Argument that having both `then` and `done` is too much for programmer's head is quite controversial to me. Both methods serve two different use cases, and both are equally important.  
Also looking at other API examples you can see it's not true. e.g. programmers somehow manage well both `map` and `forEach` on arrays. I haven't seen complains that we should remove `forEach` and just use `map`, as it's confusing and implementation might automatically _not create_ returned arrays if it detects it's not taken.

**Comment by annevk** [comment]
Per http://lists.w3.org/Archives/Public/public-script-coord/2013JulSep/0571.html we should reconsider this.

**Comment by domenic** [comment]
Might be time for @wycats to step in, as he's one of the most prominent anti-`done` people.

**Comment by wycats** [comment]
I don't feel like arguing with BE. I strongly believe that this is a scenario solving mistake that we are hastily making as promises have gained popularity.

I have come to believe that the fact that done is needed at all is a fatal flaw in the current specification. Swallowing honest-to-god exceptions (TypeError, ReferenceError) as though they were promise rejections was just a mistake.

At minimum, it should be possible to configure a promise library to bubble out exceptions, for debugging.

I would like it if we didn't repeat this mistake with DOMPromises, and avoid compounding the mistake by complicating the API.

**Comment by mhofman** [comment]
I've been reading and thinking a lot about this error / rejection handling issue in recent days and I've become convinced that the answer is in keeping Promise's semantics simple so that it mirrors synchronous calls.

In that sense, `then` is used to execute another operation after the previous one completes and `catch` is used to handle an operation failing. For the same reason, I would also like to see `finally` as I expressed before in #18 but I guess I could live without it for now.

Discriminating on the failure's type is a dangerous business since I doubt we can be sure a TypeError / ReferenceError is always due to a promise's consumer. For example, what if such an error had actually bubbled from inside a library the consumer is using? Should the consumer's asynchronous flow be interrupted even if he had a catch in place for handling such errors?

I think the burden of making sure asynchronous operations are as easy to debug as synchronous ones lies with the browser / JavaScript engine. Logging unhandled rejections is the most obvious aspect of this but there are plenty of other things that can be done to improve the developer's experience. For example engines could provide the asynchronous call stack when using promises. This is a feature available in the newest Visual Studio 2013 and it supports WinJS promises (http://blogs.msdn.com/b/visualstudioalm/archive/2013/07/01/debugging-asynchronous-code-in-visual-studio-2013-call-stack-enhancements.aspx)

Asking the developer to finish their asynchronous flow with a `done` creates the potential for them to forget to do so just as much as they now forget to finish it with a `catch`.

**Comment by domenic** [comment]
I agree with everything @mhofman says, with the caveat that I am not sure what the exact solution would be for consoles (i.e. non-interactive debuggers). I would like a solution that does not affect the authoring-time experience, though.

**Comment by wycats** [comment]
The `try/catch` that wraps the success and failure handlers is catching errors that by definition were triggered by consumer code.

**Comment by annevk** [comment]
@erights, @BrendanEich, way in maybe? Still feels like something we can address later.

**Comment by mhofman** [comment]
Yes, those errors are triggered by JavaScript code, but they don't automatically originate in the code written by the developer (consumer of the promise). My point was that a developer might want to handle such errors himself.

Regarding console type environments, couldn't the unhandled rejection be considered as a failure? I'm not sure of the consequences but for example right now an error in a setTimeout in node.js will cause the program to terminate.

**Comment by wycats** [comment]
@mhofman yep. Bottom line, many these issues are best handled by debugging environments, but as promises have quickly become more popular, and debuggers haven't caught up, people are trying to slam in the `.done()` solution to PATCH IT NOW.

I'm working on the Ember Promises debugger (for the Ember Inspector) and have talked with Dave Camp a bunch about plans for the Mozilla debugger, and making API decisions right now (before even the first iteration of debugger tools for promises have shipped) strikes me as very short-sighted.

That's why my proposals are almost all modes that you can turn on in debugging as a stopgap, and not permanent changes to the core spec.

**Comment by Nathan-Wall** [comment]
As someone who spent a long time thinking throw -> reject was a mistake in promise design, I have done a 180 fairly recently and wholeheartedly support the no-`done`-patch sentiments in this thread.  There was a time when browsers didn't have good debuggers for _regular errors_; that's just where we happen to be right now with promises.  Adding `done` to the language just so that you can see certain errors in modern debuggers would be akin to teaching people to `alert` all errors in 2003 instead of using `throw` because they were practically invisible (or impossible to get the right message/line numbers from) back then.  I mean, `alert` was what you had to use when you were debugging, but it would have been bad design to use it in general purpose code because it'd be impossible to recover from it.  Since an error thrown in `done` is impossible to recover from, it would be bad design to use, and shouldn't be part of the language.

Like Anne said, implementations that don't provide good debuggers for promises won't be used to debug problems in async control flow (or they'll be patched with tools like Firebug).

**Comment by ForbesLindesay** [comment]
I don't want `.done` for debugging purposes (that's a useful side effect).  I want it for two reasons:
1. It lets me be explicit about what I want to happen (just like I still use `.forEach` whenever I don't want to make use of the result I'd get via `.map`.
2. It is an **extremely** useful tool for explaining `.then`.  I just gave a talk (2 days ago) on promises and described `.done` first, and then described `.then` as a helper method for `.done` when you need a new promise as the result.  Everyone in that room instantly understood how `.then` works and what it does.  Explaining `.then` first **doesn't work**.  Explaining `.map` without first explaining `.forEach` **doesn't work**.

We need `.done` so we can explain what `.then` does.

**Comment by ghost** [comment]
@Nathan-Wall Your analogy to primordial JS debugging breaks down when you realize that the root problem is that under the current paradigm, there is no possible way to distinguish an "unhandled" rejection from a program error.  "Monitoring" tools attempt to sidestep this difficulty by placing the burden of distinction upon the programmer at the time of debugging.  But programming languages should be designed so that the programmer can state his or her intentions **in the source code itself**.

Some random thoughts:
- One cannot ship an API that is not debuggable with built-in features.
- One cannot ship an API that is debuggable only with features that have not been built yet, or have not been fully prototyped and tested.
- There are significant problems with `done`, as stated by others in this issue.
- Garbage collection of a rejected promise which has not been "handled" does not necessarily indicate a program error.
- The best solution to this issue would likely require us to rethink our paradigm to some degree.

**Comment by annevk** [comment]
A lot of those seem false. We ship a lot of APIs that for debugging require non built-in features. E.g. lots of network APIs don't expose sufficient information for easy debugging (due to not wanting to reveal that information).

**Comment by medikoo** [comment]
> Asking the developer to finish their asynchronous flow with a done creates the potential for them to forget to do so just as much as they now forget to finish it with a catch.

@mhofman It's not like that. The problem is that you treat `done` as a supplement of `then` which is logically wrong **. `then` is a mapping function that developer should choose over `done` and not otherwise.**
`done` should be treated as first choice function. It's the fundamental, straightforward way to access resolved value. It is `then` that someone may forgot about not `done`.

So, current state of the API provides us with brilliantly designed mapping function, but we miss one that allows us to process/access resolved values natural way, and agenda is to force user to use `then` for that, and fix its flaws with not natural, implementation dependent, mechanism that can't be implemented with plain JavaScript. Doesn't it give you notion there's some logical flaw here?

As a developer I expect that the API provides me with fine error handling control, not specifying `done` you're taking that away and I think there's no other JavaScript API that does that.

**Comment by domenic** [comment]
@medikoo, I get that you have different ideas for how you would design and use a promise API. You fail to recognize that not everyone wants to use your library's paradigm.

**Comment by medikoo** [comment]
@domenic this has nothing to do with my library, it's implementation of `done` just follows others. It'll be also great if you'd have more open and constructive approach, your negative attitude doesn't help.

**Comment by ghost** [comment]
@annevk True, to a certain extent.  The difference, to me, is that when debugging this:

```
function someFunction() { ... }
promise.then(val => someFunctoin(val));
```

I need _immediate_ feedback.  I don't want to write boilerplate (i.e. `done`, or `setTimeout`) and I certainly don't want to tab over to a promise monitoring panel to figure out why nothing is happening.  This is a program error.  The programming environment should be able to figure out when I've made this kind of error and notify me accordingly.

I think punting on `done` is understandable, but it may be that by punting on `done` we assure it's victory.  As @erights has said, a warty `done` still looks better that the proposed alternatives.  And that has been the state of things for years.  This indicates to me that something is missing from our current design with respect to error handling.

Is that missing something GC?  I don't know, but I think it would be worth our time to explore the alternatives and implications.

**Comment by annevk** [comment]
The error there is the typo in someFunction, right?

I don't understand how by punting on done we assure its victory. What strategy other than the implementation layer and something akin to done are there?

**Comment by mhofman** [comment]
@zenparsing Unfortunately the behavior you're asking for is not possible with JavaScript. Consider a similar example without the use of promises:

``` javascript
function someFunction() { ... }
setTimeout(() => someFunctoin(123), 0);
```

The ReferenceError will only be thrown when the function passed to `setTimeout` is executed, not when `setTimeout` is called. Today, to figure out there is a typo here, you need the function to be executed and then look at your console. Pormise monitoring tools would offer the same kind of error reporting.

The only way to fail at parsing in this case would be to disallow functions dynamically added to the global from being executed.

**Comment by ForbesLindesay** [comment]
1. `forEach` is the basic way of getting values out of an array, not `map`
2. `done` should be the basic way of getting a value out of a promise, not `then`

I don't see how you can dispute the second statement without disputing the first.  Does anyone dispute the first statement?  Does anyone see a reason to dispute the second without disputing the first?

**Comment by ForbesLindesay** [comment]
That doesn't mean that `map` won't be more popular than `forEach` or that `then` won't be more popular than `done`, but that wouldn't be an argument not to have `done`, unless you also think we shouldn't have had `forEach`.

**Comment by domenic** [comment]
I don't think the promise <-> array analogy holds up very well under scrutiny.

**Comment by medikoo** [comment]
@domenic can you be more constructive, do you have any valid arguments?

**Comment by annevk** [comment]
It's not about whether or not to return a new promise. It's about reporting errors. The forEach/map analogy has nothing on that.

**Comment by ghost** [comment]
@mhofman I meant the same ReferenceError that `setTimeout` would report on the console.  There's no reason why I should have to add boilerplate (`done`) or tab between two different views (the console and the promise monitor) in order to see that error.  It is a programming error and should be presented to the programmer in the same way as other programming errors.

**Comment by medikoo** [comment]
@annevk 
It's not only about that, thing that `promise.then` creates another promise that we totally don't need, affects for performance and is not clean. Also the reason of proposed error reporting is mapping feature of `then`, this comes out of it.

You want us to deal with all those side effects of `map` on each value access. I really wouldn't like to be forced to use`setTimeout` with native API's to be able to handle errors normal way. It's bad idea to leave API in such state.

**Comment by domenic** [comment]
`promise.then` will not create a new promise in any reasonable optimizing JIT, unless its return value is actually consumed.

**Comment by medikoo** [comment]
@domenic it's weird thing to provide imperfect solution and fix its issues that way. It's another mess to handle for compiler. Also mind that such thing can't be implemented in plain JavaScript (no polyfill possible).

**Comment by ghost** [comment]
@annevk I'll try to explain...

First, let's forget about the browser and just consider the server.  The appropriate thing for a console (Node) application to do when it encounters an unhandled error is to crash the program, reporting some error information.  You don't fire up a monitor.

So the only way that the current promises spec will work in that environment is to include boilerplate (`done`), or to rely on the GC to deduce that an unhandled rejection will never be handled, and therefore constitutes a program error.

To my knowledge, we've not fully validated the GC approach.  If it turns out that we can't rely on the GC, then the only other option for that environment is `done`.

The force which is pushing us toward the GC or `done` is the fact that, as currently specified, one may attach a listener to a rejected promise in an _arbitrarily distant future_.  This is the property which is wreaking havoc with our ability to report simple programming errors.  Is this property required?

**Comment by annevk** [comment]
Let's make that more concrete.

The suggestion would be that in PropagateToDerived(p) if p.[[Derived]] is empty and promise.[[Reason]] is set, you'd propagate it out of the promise? (E.g. window.onerror, console, ...)

**Comment by wycats** [comment]
@annevk I implement a similar strategy to this in RSVP (via `RSVP.configure('onerror')`). Unfortunately, it makes for a lot of noise in cases where people are intentionally trying to attach error handlers later.

I consider it worth the trouble for debugging, which is why I offer the option. I've had good luck printing error messages to the console in my implementations of the RSVP `onerror` hook only if the original error has a `stack` (i.e. it's a real error), but clearly these are only heuristics that can be overly noisy or overly silent depending on the use-case.

fwiw: The fact that people don't typically use debugger tools in node yet doesn't mean that we should assume node will never have a popular, good debugger environment.

**Comment by annevk** [comment]
Okay, so @wycats' comment indicates that even if we did want that, it'd have to be opt-in, and thus could be added at a later stage. So we have three strategies: 1) done() 2) GC/engine magic 3) a flip to twiddle PropagateToDerived handling. All three can be deployed after-the-fact. Unless there's anything else, I recommend we close this.

**Comment by medikoo** [comment]
> The suggestion would be that in PropagateToDerived(p) if p.[[Derived]] is empty and promise.[[Reason]] is set, you'd propagate it out of the promise? (E.g. window.onerror, console, ...)

As @wycats noted it is imperfect solution, and I think it shouldn't be taken into account

>  So we have three strategies: 1) done() 2) GC/engine magic 3) a flip to twiddle PropagateToDerived handling. All three can be deployed after-the-fact

If promise implementation would implement in full given API but would not provide any of above, would you consider it as complete and fine to use?

Other important thing we need to realize, if some implementations will do 1), and others will do 2) we will deal with promises implementations that would conform to same API but would need to be used in different way.
That means, code written for A would not work as expected in B, and other way, and we're not talking about optional custom features, it's about errors handling which is mandatory thing to solve. We're introducing disruption.

**Comment by annevk** [comment]
That's bullshit. Implementations are not allowed to deviate from the standard. The implementers know at this point in time their only option is 2. We can then evaluate later on whether we actually want 1, 2, or 3 or a combination. It's not about complete, software never is, it's about what we can ship right now and start playing without creating problems going forward.

**Comment by medikoo** [comment]
> That's bullshit. Implementations are not allowed to deviate from the standard.

It's exactly my point. We need to specify that, so we have one solution and not few different and incompatible (lack of specification leads to deviations).

> The implementers know at this point in time their only option is 2. 

Is it specified somewhere, or just decided behind the curtains? Implementers of which engines agreed on that?

>  It's not about complete, software never is

We're talking about mandatory functionality that will make this API usable. It's `Promise.prototype.catch` that can be ignored and added at any time later, but not `Promise.prototype.done`.

**Comment by ForbesLindesay** [comment]
I keep failing to get people to see my point.  The analogy between array and promise holds up brilliantly in this case.  I'm not arguing for something to help error handling.  I'm not arguing for something to help performance by not creating an additional promise (I've implemented `done` in terms of `then` for my promise implementation).  **I'm arguing for a method that lets me say what I mean.**  I want to express my intent when writing the function.  That is to say, I want to express the difference between "Get the result of this promise" and "Transform this promise with an additional operation".

**Comment by ForbesLindesay** [comment]
Even if we could solve the halting problem and always detect promises that would never be handled, I'd still want `done` so I could express my intent.

**Comment by juandopazo** [comment]
@ForbesLindesay FWIW it's not just about an analogy with arrays. Other languages have both features. For example Scala has `onComplete` and `map`.

**Comment by erights** [comment]
> Even if we could solve the halting problem and always detect promises that would never be handled, I'd still want done so I could express my intent.

@ForbesLindesay FWIW, I expect to want it too. But it would be a compatible addition, and so can come in standards that build on this one.

> It's [P]romise.prototype.catch that can be ignored and added at any time later, but not [P]romise.prototype.done.

@medikoo I agree with the first part of this statement. I also agree that the argument for including .catch is weaker than that for .done, so including .catch and not .done seems mistaken. If you want to start an issue to remove .catch from this repository, I would be in favor. .catch is also something that would be a compatible addition later and so need not happen in this round.

**Comment by medikoo** [comment]
@erights ~~To me `catch` is redundant to `then` not `done`, it's actually just sugar that can easily be implemented on top of `then`~~ _(edit: sorry misunderstood your point at first)_

I don't mind `catch` in the API, it's just surprising to see such optional function being added and `done` being ignored.

**Comment by ForbesLindesay** [comment]
I could cope with not adding `.done` now.  My fear is that not having it from day one will make it a lot harder to teach people to use promises (in turn leading to more fear and hatred of them, which there is already far too much of).  I'm also concerned that most people are not saying "We can add this later, for now lets just get something out the door" they're saying "This won't be needed once the development tools have caught up, lets never add it."  I feel like there might not be another opportunity as good as the current one when it comes to getting this implemented correctly.

**Comment by erights** [comment]
Please see https://mail.mozilla.org/pipermail/es-discuss/2013-August/032486.html
I take the purpose of this repository to be guidance for item 2:

```
2) What should DOM do quickly [with] tc39's quick output?
```

which refers back to

```
1) What should tc39 do quickly, to unblock the DOM's need for promises and avoid a design fork?
```

At the upcoming tc39 meeting, I will thus hope to see tc39 adopt a compatible superset (item 1) of the draft standard specified by this repository, enabling DOM to adopt the draft standard specified by this repository itself as the subset demanded by their immediate needs (item 2).

The superset I will advocate will include .flatMap and .of, per the AP2 agreement. And it will include .done.  Both because I already thought .done would be necessary, and because I find the arguments in this thread to be compelling new reasons for wanting .done. However I still don't think .done belongs in item 2, and so not in this repository.

As for item 3, we will be able to decide based on experience we don't yet have. This is good.

**Comment by medikoo** [comment]
@ForbesLindesay

>  I'm also concerned that most people are not saying "We can add this later, for now lets just get something out the door" they're saying "This won't be needed once the development tools have caught up, lets never add it." 

This is my big concern as well, and not specifying (or assuring existence of) `done` opens pandora's box of issues.

@erights sounds reasonable. I just can hope that indeed `done` will land in DOM and through that all will follow.

**Comment by ghost** [comment]
@wycats I'm curious.  Can you provide an example of a programmer "intentionally trying to attach error handlers later"?

Thanks!

**Comment by juandopazo** [comment]
@zenparsing preloading stuff into a cache?

**Comment by erights** [comment]
@zenparsing http://research.google.com/pubs/pub40673.html
The mint/bank deposit method returns a promise for whether the transfer of money succeeds or fails. Later or earlier, or even on a different machine, the escrow exchange agent does a Q.all on several of these promises, which register onReject handlers on them.

In my experience, this kind of delayed registration is quite common, and is one of the reason that promises are so convenient.

**Comment by domenic** [comment]
OK. So `done` is outside the scope of this repository; closing this issue. But to be clear, as @erights states, it's not off the table entirely for ES7 promises. We can put it in the bucket of things, like `finally`, that at least some people want ASAP as additions to the DOM promises subset this repository is designed to codify.

**Comment by spion** [comment]
I don't get it. Although late registration is common it still looks like a very special case to me. The vast majority of cases I've encountered don't involve sending promises over the wire and therefore don't seem to benefit from it. 

On the other hand, it seems _very_ common to forget the need to use `.done` at the end of a longer promise chain and end up with bugs that leave no traces and take forever to track.

Wouldn't it make more sense to have `promise.undone()` instead to indicate that the promise will be handled later? Or am I missing something obvious?

**Comment by medikoo** [comment]
@spion you're missing something obvious. Spec in current shape is missing a function that provides access to resolved value without any side effects, `done` is the one.

The best way to get is to use promises in real projects and not just theorize about it, but following above discussion should also be helpful.

**Comment by petkaantonov** [comment]
In your example, exceptions could have been exposed even if you had written:

``` js
promise.then(function (value) {
    // Process value
}, function (err) {
    // Handle error
});
```

Because it is possible for a promises library to detect that you don't have an error handler. So it is just common sense to report it or even throw it in the process.

Problem is that an error handler attachment could be delayed. And that's Spion's point: Why should everyday common usage be punished by requiring some obscure easy-to-forget method calls like `.done()` just because there is a possibility to attach late handlers? We should instead require those late handlers to call obscure methods, not the everyday average developers.

Also, you should not worry about promise object cost like that. You happily create 3 functions (1 of which is closure) regardless of path taken when your code is called which is already worth 5-15 promise objects depending on implementation so mentioning perf is not a strong point.

**Comment by wycats** [comment]
If you look upthread, you'll see that I proposed this a few times. The response argument, which I think I agree with after trying it, is that it's not always clear to a piece of code that receives a promise whether the error handler will be attached synchronously or asynchronously.

When you start building a system that abstracts asynchrony (especially one that forces asynchrony for consistency, like this one), it becomes very easy to lose track of whether some part of you code is attaching things synchronously or not. Users will then receive unexpected exceptions, without a clear understanding of the root cause.

Sadly, one's perspective on this problem changes over time from the beginning (when asynchronous attachment seems remote) to later on (when opting out of synchronous attachment seems like a huge burden).

**Comment by petkaantonov** [comment]
> Users will then receive unexpected exceptions, without a clear understanding of the root cause.

The less extreme alternative I gave was just to report it, e.g. write the stack on stderr as possibly unhandled exception. You could then investigate it and either call undone() to signal that this has late handlers or fix the bug.

> Sadly, one's perspective on this problem changes over time from the beginning (when asynchronous attachment seems remote) to later on (when opting out of synchronous attachment seems like a huge burden).

Yes, but the argument is still that it's better to force .undone() on people who do late attachment than to force done() on everyone, because late attachment is the less common case. 

Unless one is mixing async and promise code instead of wrapping/promisification then as I see it, it requires deliberate setup to attach delayed handlers. I'd really like to see a code example where a delayed handler is attached inadvertently that doesn't involve remote promises or mixing paradigms.

**Comment by medikoo** [comment]
@petkaantonov It's not possible to fix side effects of `then` in bulletproof way in plain JS implementation, and in native implementation it requires specific optimizations and some custom error handling mechanism that no other API has seen so far.

What is obscure it's how we want to fix the issues that we introduce by not providing `done`. 

Argument that `then` and `done` are too much for average developer's head is totally flawed:
- We already deal with similar concept in Array API (`forEach` and `map`), I don't recall anyone stating that's too confusing.
- If developer is not familiar with propagation of promises, he/she will have big problems with lone `then` already. There's [when.js](https://github.com/cujojs/when) library which tries to provide just `then` and solve unhandled errors with some kind of monitor, and still there's a lot of misunderstanding, just browse the issues of this project to see what it turns into.

`then` and `done` are different methods that serve two different use cases, both are mandatory for promise implementation, and this statement I base on my two years experience of working with promises in real world projects.

**Comment by petkaantonov** [comment]
Well, let's just say we have very different definitions of ["average developer"](http://stackoverflow.com/questions/tagged/javascript) if you think they even use `.forEach` and `.map`. :smiley: I will check the when.js issue tracker but last time I checked the library doesn't provide long stack traces so I can see already how it would be less useful.

**Comment by ghost** [comment]
@spion Thank for bringing this up - I believe that your intuition is correct.  I am preparing a post for es-discuss which will address this and other issues with the current Promise spec.

**Comment by spion** [comment]
I came up with an example that makes me wonder if I'm correct:

``` js
function fetchCachedOrRemote(item, remoteDataStorePromise) { 
    return cache.get(item).then(function(result) { 
        if (result) 
            return result; 
        else 
            return remoteDataStorePromise.then(function(store) { 
                return store.fetch(result).finally(cache.put); 
            });
    });
}
```

Not waiting for the remoteStore promise to resolve is not only convenient but also potentially much faster.

However, in this example its still beneficial to attach an error handler to the remoteStorePromise (e.g. to handle reconnection). 

But I wasn't trying to make a case against `.done`, I was just looking for a simple example that demonstrates why attaching error handlers asynchronously is clearly more convenient. And I see how medikoo's argument is conceptually sound: `.then` creates a new promise, `.done` doesn't - so if you don't want a new promise, don't use `.then`

**Comment by erights** [comment]
>  a simple example that demonstrates why attaching error handlers asynchronously is clearly more convenient

See http://research.google.com/pubs/pub40673.html
Even when used non-distributed, the patterns in this code do delayed registration.

**Comment by ghost** [comment]
@erights That code, although dense, is a great illustration of delayed registration patterns.  I agree that delayed registration is an important feature, but the question is:  What is the best way to support both delayed registration and also proper and timely identification of program errors?

`Done` does the job, but at the cost of (a) bifurcating the callback API, and (b) requiring that the user add `done` at the correct places.

More thoughts to follow...

**Comment by ghost** [comment]
A concrete proposal:

Conceptualize promises as nodes within a forest of trees.  Root nodes are created with the promise constructor:

``` js
var root = new Promise(...);
```

Child nodes are created with `then`:

``` js
var child = p.then(...); // "child" is a child of "p"
```

Any leaf node which is rejected is clearly an unhandled error because there is no child for the error to flow into.  The problem is: how do we know whether a given node is a leaf or an intermediate node whose children have not yet been added?

Under our current error handling model, all nodes are assumed to be non-leaves.  That is, we assume that a child may be attached at any event in the future to any node.  We are therefore unable to determine if any rejection is an unhandled error.  Essentially, `done` provides us with a means of explicitly marking a node as a terminating leaf and thereby restoring that ability.

What if we reversed the assumption?

Let's assume that every node is a leaf, until a child has been added to it.  More specifically:
- In PropagateToDerived(p), if p.[[Derived]] is empty and p.[[Reason]] is set, then queue a task to run **immediately after the microtask queue is flushed** which runs the following:
  - if p.[[Derived]] is empty and p.[[Reason]] is set, then _propagate an unhandled exception in a platform-specific manner_. 

Additionally, add a new convenience method to the Promise prototype to explictly make a node a non-leaf:

``` js
Promise.prototype.delayed = function() {
  this.then(null, reason => reason);
  return this;
};
```

This method (which needs a better name) can be added at a later date.  For the present it can be provided by a helper library.  Coding patterns which depend upon delayed registration would need to invoke such a method.

**Comment by Nathan-Wall** [comment]
> What is the best way to support both delayed registration and also proper and timely identification of program errors?

This answer seems easy and clear to me:  Proper debugging tools for uncaught `then`s.

**Comment by ghost** [comment]
@Nathan-Wall That sounds cool, but:
- It presupposes an interactive debugging environment.  That's fine for the web, but do you want to make such an assumption for, say, Node?
- If you are making heavy use of delayed registration, then presumably you'll have lots of things coming and going from your promise debugging tool.  There will be noise that you'll have to filter (perhaps using some heuristic algorithm).
- I would think that ideally, "program errors", whether coming from sync or async patterns, should be reported in the same way.  Otherwise, aren't we requiring the user to "switch views" constantly?

**Comment by wycats** [comment]
@zenparsing I proposed precisely this a while ago (with the worse, tongue-in-cheek name `undone`).

The problem is: who exactly in an abstraction should call it? The delayed attachment mostly happens as a side effect of a lot of async, rather than an intentional choice to delay attachment.

In other words, there isn't really a point where you think "I'm going to delay attachment now", it's something that just starts happening as the system grows.

**Comment by ghost** [comment]
@wycats What convinced you of that conclusion?  Can you provide an example?

**Comment by medikoo** [comment]
@zenparsing 

> Done does the job, but at the cost of (a) bifurcating the callback API, and (b) requiring that the user add done at the correct places.

Don't you see that your proposal is much more obscure and makes things much worse?

Many say `done` and `then` is too much for dumb developer (which in my opinion is nonsense). I ask do you expect that your proposal would be much easier for developers to adapt?

Do you think developer will less likely forget to run `delayed` which now you _require to add_ instead of `done`? and mind that `done` solves many more problems than one you're trying to solve in cumbersome way here.

There's _null_ cost with `done` just benefits, and please realize `done` is not an add-on, if you want to look that way, it's `then` that is add-on to `done` not otherwise. `done` is the most atomic, generic one, that makes one thing right without any side effects. All API (including `then`) could be build on top of `done`, but not on `then` which does a bit more.

We're making attempts to reverse cost of `then` with crazy complicated, imperfect solutions and fail to realize that all that needs to be done is to provide more atomic `done`. It's not what future users of promises expect from us.

**Comment by petkaantonov** [comment]
`delayed` is not required in place of `done`. It is not required at all. It is only when you do something like pass around promises as if they were event emitters that you would potentially need to call `delayed`. I can agree that telling in such a situation whether async or sync attachment is happening is hard but I disagree that this situation is common in the first place - hell, all the examples everywhere (except the paper) use promises for their sync parallel, not as event emitters.

And even if you forget `delayed` all what happens is that you may get a message about potentially unhandled rejection. Not as bad as forgetting `.done`

**Comment by ghost** [comment]
@medikoo The thrust of your argument is that `done` / `then` are analogous to `forEach` / `map`.  I understand what you're saying, but you're back-porting this analogy onto the design, rather than building it _into_ the design.  If you were going to build it into the design, you would choose different names.  With a dual-callback API, `then` becomes an attractive nuisance because it's _such_ a nice name.  The API just wasn't designed for dual-callback.

Furthermore, my other point still stands:  the user still has to remember to add `done` in the correct places in order for errors to show up in the error console.

@petkaantonov has it right.  When you forget to add `delayed`, you get a nasty error real quick.  This is how error handling ought to work.  The programmer should not have to _ask_ the environment for an error notification.

@wycats might still have a point though, we'll have to see.  Although my first reaction is to say that as a general rule abstractions should be designed so that errors can be handled by _something_.  : )

**Comment by ForbesLindesay** [comment]
The assumption that you get an error quickly when you forget to add `delayed` is not valid.  When you forget to add delayed, it will probably work 95% of the time.  The other 5% of the time, there will be a massive error because of the race condition you create.  Since the promises are asynchronous, there's always a chance that the promise will take long enough to resolve that it gets watched before it resolves.

**This issue is not to discuss the idea of early errors when you fail to handle errors in the same tick.  If you want that, please open a separate issue.  This issue is for deciding whether we want `done`.  To be clear, making `then` throw an error that would crash an application is not an option we should be considering in this issue.  There is a very important issue that does warrant discussion, and this issue is rapidly becoming long enough to need a summary.**

**Comment by domenic** [comment]
>  If you want that, please open a separate issue.

Actually, please don't, or at least be aware that it will be closed immediately. Both `done` and any other proposed things anyone wants along these lines are outside of the scope of the TC39 September meeting consensus which this repo is meant to flesh out, and are off-topic to the real spec work going on here.

I would prefer that all this `done` rehashing take place elsewhere as well, but I don't feel like banning people, so whatever, if you guys want to use a closed issue as your own private mailing list that's fine I guess.

**Comment by ghost** [comment]
@ForbesLindesay Easy mate.  We're just exploring - for kicks. : )

Regarding your argument, I think I understand what you're saying.  Need a minute to think about it though...

**Comment by medikoo** [comment]
> If you were going to build it into the design, you would choose different names.

That's possible, but it's too late to start over with blank paper, `then` and `done` names were already coined in many libraries, and we should stick to them.

> user still has to remember to add `done` in the correct place

not to add, to use **instead** of `then` in the correct places

**Comment by ghost** [comment]
@ForbesLindesay Let's take a look at this.

Assume that I've unintentionally left out a `delay` on a promise which is supposed to be delayed.  And let's assume that the resulting error was not revealed in testing due to environmental (async) concerns.  What are the consequences?
1. For a web application, an error message will get logged to the console and window.onerror will get called.  Otherwise the application will keep running as intended.
2. For a server application, it depends upon how the application handles uncaught errors.  As a base case, the program will crash with a stack trace.  More complicated applications might log the error before attempting some kind of graceful restart.  Either way an issue will get logged and the bug will get fixed.

Since the system has no way of knowing that a rejection (which I forgot to delay) isn't in fact a program error leaving the program in an inconsistent state, then I think both consequences above are appropriate for their respective environments.

**Comment by ForbesLindesay** [comment]
@zenparsing read the entire discussion.  Debugging tools will be able to catch up and trap unhandled rejections.  They will also be able to report the bulk of the "never handled rejections" by hooking into the garbage collector at some point in the future.  Non-deterministic bugs are by far the hardest things to fix.  The way things currently are makes these bugs deterministic, which is a huge benefit.

@domenic I realize that it would be an issue that got quickly closed, but if that discussion had happened in a different thread, this could remain a useful record of the arguments for and against `done`.  Sadly it's probably too late now :(

**Comment by ghost** [comment]
@ForbesLindesay I don't quite understand what you mean when you say that the current model makes "these" bugs deterministic.  Can you elaborate?

**Comment by spion** [comment]
The current model always swallows these bugs when you forget to use `done()` but you're not using the return result of `then()`. In that way, its deterministic.

The `delayed()` model will swallow forgetting to use `delayed()` _sometimes_ - only if the promise rejection happens before attaching the rejection handler.

``` js
function fetchCachedOrRemote(item, remoteDataStorePromise) { 
    // forgot to do remoteDataStorePromise = remoteDataStorePromise.delayed();
    return cache.get(item).then(function(result) { 
        if (result) 
            return result; 
        else 
            return remoteDataStorePromise.then(function(store) { 
                return store.fetch(result).finally(cache.put); 
            }); // finally handled via return value
    });
}
```

If I forget the first line in `remoteDataStorePromise`, there are two possibilities
1. `cache.get` responds faster than remoteDataStorePromise - no error
2. `cache.get` responds slower than remoteDataStorePromise - error.

Which in some situations may lead to bugs that can only be reproduced when the sun aligns correctly with the moon (e.g. the cache server is under super-heavy load, like in production)

Attaching to the GC is the only sure way to deal with this. Since there are no more references to the promise anywhere in the program, no further handlers can be attached. Therefore the runtime can report the error (e.g. log it or throw it in node.js). Perfect.

**Comment by ghost** [comment]
@spion Thanks - this is just what we need.

First, I hardly think swallowing of program errors can be considered helpful, deterministic or not.  So we can leave that argument behind.

Second, I wouldn't be so sure that using the GC is "perfect".  First of all, the garbage collector does not give you any guarantees about whether a particular object will be collected.  (Correct me if I'm wrong, please.)  Furthermore, it is quite trivial to construct a case where the GC does not collect a rejected promise which represents a program error:  simply assign the promise to a long-lived variable.

To put it more concretely, given this (note the misspelling):

```
function F(someVar) { G(someVr); }
```

why should this:

```
fetchURL("/some/url").then(F);
```

have different error handling semantics than this:

```
var p = fetchURL("/some/url").then(F);
```

Before I go into your code example, let me point out that under the current model detecting an unhandled program error [edit:  within a finite time of the error  ; )] is undecidable (in the full Turing sense of the word)*.  This is a vast departure from stack-based error handling and it deserves exploration.

*I think this is a valid reduction: https://gist.github.com/zenparsing/7030884

**Comment by ghost** [comment]
Let's take a look at @spion 's code (reformatted somewhat):

``` javascript
function fetchCachedOrRemote(item, remoteDataStorePromise) { 
    return cache.get(item).then(result => { 
        if (result) return result;

        return remoteDataStorePromise
          .then(store => store.fetch(result))
          .finally(cache.put);
    });
}
```

It seems to me that implicit in this code is the assumption that `fetchCachedOrRemote` is not "responsible" for any errors that may be represented by `remoteDataStorePromise`.  It uses that promise, but it does not "own" it.

Also, the "responsibility" for any other promises generated in this function are transferred out to the caller.

Do you agree?

**Comment by spion** [comment]
Yes I agree - my example is lacking. Whoever created the remoteDataStore promise is ultimately responsible of handling its error conditions (e.g. retrying the connection, perhaps). Therefore my example only shows that late registration is beneficial in general, but not that there is a need for late registration of error handlers

Another important thing that I noticed is that the example in the paper posted by @erights is using promise cancellation for control flow while @zenparsing and I are thinking of handling error conditions / recovery - two very different things with different handling responsibilities.

I still haven't managed to come up with an example where delayed registration for error handling/recovery is obviously useful as opposed to early error handling. I'm currently searching for one :) If anyone with more experience using promises can come up with a simple one, I'd be really glad.

However, I now understand that `.delayed()` / `.undone()` will almost definitely **not** help - it will only result with hard to reproduce bugs.

@domenic - what would be the best place to continue this discussion - perhaps es-discuss? I feel like we're abusing this issue as well as possibly annoying the people subscribed to it...

**Comment by ghost** [comment]
> However, I now understand that .delayed() / .undone() will almost definitely not help - it will only result 
> with hard to reproduce bugs.

I agree.  As proposed, `delayed` does nothing but swallow errors.  Let's drop it from the discussion.  It doesn't bear on our present line of inquiry, anyway.

**Comment by ghost** [comment]
Clearly decidable error handling (as I proposed above) is no less powerful than undecidable error handling - one could simply "delay" any promises that are going to be handled later.  But as we've seen, that isn't a very good programming strategy.

Our task now is to come up with a reliable pattern for handling errors within the context of a decidable error handling model.

First, some terminology:
- **Up the call stack**:  In the direction of the caller of the current function.
- **Responsibility**: For a function, the state of having the duty to handle a rejection.
- **Handled promise**:  A promise which is the result of a direct call to `then` whose error callback converts all rejections into values.
- **Unhandled promise**:  All promises which are not handled.

I propose the following strategy:

**Responsibility flows in one direction only:  up the call stack.**

Or, more precisely:

Unhandled promises that flow up the call stack into a function invocation F must be either:
- pipelined (resulting in a new promise to which this rule is recursively applied),
- handled, or
- returned up the call stack to the caller of F

before the F is cleared from the call stack.

Can we think of any use cases that this rule leaves out?

**Comment by ForbesLindesay** [comment]
@zenparsing or the promise must be stored in a variable somewhere, that can be later referenced and handled.  You're trying to solve the halting problem (which is known to be impossible).  The garbage collector does a pretty decent job.  We can't "crash the app" when the unhandled rejection is noticed by the garbage collector, but we can log it and alert the developer to the error.

The reasoning for done has only a little to do with error handling, and far more to do with how purely your intent can be specified.

**Comment by medikoo** [comment]
> The reasoning for done has only a little to do with error handling, and far more to do with how purely your intent can be specified.

Actually It has a lot to do with error handling as well. It's the only function that allows us to process resolved value outside of promise specific error handling mechanism. I see it as no less important feature.

Mind also that API in current shape doesn't provide any way to expose unhandled errors (they'll be silent). Therefore it cannot be perceived as valid for a complete implementation. Adding `done` would solve that shortcoming.

**Comment by ghost** [comment]
@ForbesLindesay I'm certainly not trying to solve the halting problem!  As I've laid out, the current error handling model is undecidable.  That is, the problem of determining whether a particular error has a handler is undecidable.  In my opinion, this is a mistake.  

On the other hand, under the proposal at https://github.com/domenic/promises-unwrapping/issues/19#issuecomment-26254979 the problem _is_ decidable.

The question becomes, do we lose anything with decidable error handling?  I don't think we do, but I'm trying to work my way through that question.

**Comment by ForbesLindesay** [comment]
You can either:
1. Loose deterministic behavior (this isn't going to be accepted)
2. Throw/Log on garbage collected unhandled rejections (this won't catch every possible case, but gets most of the way)
3. Add a `.done` method (this allows pure expression of intent, will catch any case where it's added, but can be accidentally forgotten).
4. Solve the halting problem (this is the only thing that catches everything.  It's actually the only thing that catches more than the garbage collector without people explicitly adding then.)

1 is a non-starter.  2 and 3 are both simple and extremely useful.  4 is proven to be impossible.

**Comment by ghost** [comment]
The question I'm pursuing is, can we "do" promises in such a way that error handling is both deterministic (to use your terminology) and decidable.

If we adhere to the rule at https://github.com/domenic/promises-unwrapping/issues/19#issuecomment-26606667 , do you agree that error handling is deterministic?  Or am I misinterpreting your meaning?

**Comment by Nathan-Wall** [comment]
@ForbesLindesay, there's another version of 2 which is to log unhandled rejections and unlog them when they get handled, so that it's not dependent on the garbage collector... maybe the debug tools could permit you to filter by rejections that had been garbage collected for a better view, but non-collected unhandled rejections could be made visible.

**Comment by petkaantonov** [comment]
How is `.done()` (without arguments) not just a `.catch(thrower)`.. so just do:

``` js
Promise.prototype.done = function() {
    this.catch(thrower);
};
```

**Comment by juandopazo** [comment]
If your thrower function is making sure the exception gets thrown, the only difference is that `done` shouldn't return a new promise becaus it's not a mapping operation.

**Comment by petkaantonov** [comment]
Yes I removed the `return`

**Comment by medikoo** [comment]
@petkaantonov `catch` is just sugar over `then`, and you should also explain what's `thrower` in your sample, as I assume it's nothing provided by the standard.

`promise.done(processValue)` vs `promise.then(processValue).catch(thrower)`, latter ugly as hell, produces 2 obsolete promises (not needed for anything), and error is exposed next tick later than in case of `done`.

**Comment by petkaantonov** [comment]
My point was that `.done` without arguments is trivially implementable even if it's not provided out of the box. Also what difference does it make if your server crashes with a 3 micro second "delay"?

``` js
function thrower(err) {
    process.nextTick(function(){
        throw err;
    });
}
```

**Comment by medikoo** [comment]
@petkaantonov Issues are
1. Cleanness of code, need to put `catch(thrower)` at end of each promise chain. We really don't want to do that, and mind we need to provide `thrower` ourselves.
2. Obsolete promises created for no valid reason, that affect performance (speaking to performance purist ;-)
3. Next tick error resolution is just another dirty aspect of it.

**Comment by petkaantonov** [comment]
You _can_ just put `.done()` since you can just add it to prototype (aren't they a wonderful thing). Although `.catch(thrower)` actually says what it does but whatever.

A chain of promises requires creation of closures due to unutilized `this` so the created promise object is literally a drop in the ocean.

**Comment by medikoo** [comment]
> You can just put .done() since you can just add it to prototype (aren't they a wonderful thing).

That's not possible. You can cleanly implement `then` over `done` but not the other way. You can try to mimic `done` behavior on top of `then` as you suggest, but that's far from perfect, it's a poor man's solution.

**Comment by ForbesLindesay** [comment]
Yes, user's can add their own `.done` method, no that does not mean it should not be in the spec.  It has already been well proven in user space.  We know it is useful, so it is ready to be specified and then implemented.

**Comment by jakearchibald** [comment]
Here's the V8 ticket for logging unhandled rejections https://code.google.com/p/v8/issues/detail?id=3093.

The current plan is to console.error rejections to promises that don't have a rejection handler on them specifically or further down the chain.

So this will console.error:

``` js
Promise.reject(Error("BROKE"));
```

This will not:

``` js
Promise.reject(Error("BROKE")).catch(function() {
  return "Nah, it's ok";
});
```

However, this will console.error (a false positive):

``` js
var p = Promise.reject(Error("BROKE"));

setTimeout(function() {
  p.catch(function() {
    return "Nah, it's ok";
  });
}, 500);
```

We're hoping the above case is rare enough.

**Comment by medikoo** [comment]
@jakearchibald there are cases like that, and they're not that rare (especially in node environment). So you'll definitely have false positives around.
Such solution was already implemented by some libraries a while ago. [RSVP.js](https://github.com/tildeio/rsvp.js) has it and [When.js](https://github.com/cujojs/when) (which added `done` to it's API recently) also. Check their issues, it was already proven as problematic.

Aside of general flaw that this solution has, other question is: What about other JS engines, and I'm not speaking of just browsers. Any implementation that would silent unhandled exceptions is doomed. I hope we all agree in that.

**Comment by juandopazo** [comment]
@jakearchibald `Promise.reject` logging errors becomes too noisy very fast. That particular case should not log errors. This should though `var p = new Promise((resolve, reject) => reject(foo));`.

In YUI's implementation I'm specifically circumventing the unhandled errors logic in `Promise.reject`: https://github.com/juandopazo/yui-promise/blob/master/promise.js#L248-L259.

**Comment by jakearchibald** [comment]
Understood, will dig out the issues for other libraries and take them to the V8 team.

Sounds like debugging is in a really bad place.
- `done` is seen to put too much responsibility on the developer
- `console.error` at GC time is too late, as that may not happen until unload/exit
- `console.error` for rejected promises without a catch in the future chain is too soon, as it's not uncommon to attach  catches later

So… what then?

I was convinced by this thread but it's feeling like `done` is the least-worst option. It's the developer saying "I'm not handling errors", which maps to not having try/catch, or `try { ... } catch(e) { throw e; }` in sync code.

**Comment by domenic** [comment]
The solution is pretty simple. A live list of currently-unhandled errors. One of the great advantages of browsers implementing promises is that they can accomplish this in their tooling, whereas until now library authors have had to use hacks like e.g. logging an array and counting on the developer tools to keep it live updating.

**Comment by juandopazo** [comment]
Honestly, to me `done` vs tools is a false dichotomy. We can do both. In A+ having the smallest API surface was crucial to have less to agree on. But ES6 is already introducing new methods, so I see no downside in also including `done`.

**Comment by medikoo** [comment]
> `done` is seen to put too much responsibility on the developer

Anti `done` attitude takes it's source from _"it would be beautiful to have just `then`"_ way of thinking, and not from: _"I've tried `done` and it's too confusing to have it aside of `then`"_.

Let's ask people that rely on `done` in everyday work with promises, whether they find it problematic.

I've actually came from _"only `then`"_ way of thinking to _"`done` (as first choice), and `then` when transformation is needed"_ over year ago and it was just positive, never looked back.

It's also important to note that in most cases `done` is replacement for `then` and _not_ addon that you have to and may forgot to add at the end of chain.

**Comment by jakearchibald** [comment]
> The solution is pretty simple. A live list of currently-unhandled errors

But that's something the developer has to ask for, whereas a logged error is pushed at them. Also, this live list is even less accessible in a non-graphical environment like node.js.

The list thing (which I'd love to see in the style of a git graph) will happen, but is it enough?

**Comment by spion** [comment]
Here is a summary of solutions and their shortcomings:

### Using .done()

Pros: 
- unambiguous, clarifies intent.
- no problems with asynchronous error handling

Cons: 
- Making a mistake (not using it instead of .then when you should) causes things to silently fail. 
- A lint tool is necessary to expose such mistakes
- Limited reporting when doing async error handling, e.g.
  
  ``` js
  var promise = Promise.reject(Error("First"));
  
  Promise.reject(Error("Second")).done(function() {
   promise.catch(function(e) { /* handled */ }).done();
  });
  ```

Only the "Second" error will be reported.

### Logging posibly unhandled rejections

Pros: 
- works very well in most cases
- never fails silently if you make a mistake (forget to do something)

Cons: 
- wont always warn you if you asynchronously attach error handlers (you get a race - depending on the speed of operations, you may or may not be warned)

Note: Indeed, some cases are easier to write with asynchronous error handler attachment. However, its likely that all of them can be rewritten - see [rsvp No 196](https://github.com/tildeio/rsvp.js/issues/196)

Asynchronously attaching error handlers may lead to other hazards (what if the asynchronous operation that does the attaching fails? Very bad).

### Live list of unhandled rejections

Pros: 
- you get to check current unhandled rejections instead of getting a log of false positives

Cons:
- browser-only
- potential performance issues if the list gets long
- a server-side implementation is useless - even if you write the UI, the more you use asynchronous error handling, the larger (and more useless) this list will become, as it will potentially contain a lot of errors that will soon be handled.

### Special GC strategy

Pros:
- no false positives
- no possibility to make a mistake

Cons:
- complex - requires separate treatment of promises by GC to avoid them ending up in the old space
- wont work if you're still keeping a reference to the failed promise for some reason

**Comment by domenic** [comment]
> But that's something the developer has to ask for, whereas a logged error is pushed at them.

I don't understand this sentence. You open your developer tools and click "console": you see console.errors. You open your developer tools and click "promises": you see unhandled rejections. Probably you proxy some unhandled rejections over to the console, like you do XHRs---but you need to _remove_ them from the console once they become handled.

**Comment by stefanpenner** [comment]
great summary @spion.

My gut feeling is recommended best practices would combine several strategies which would yield the best-possible outcome. 

Maybe someone can find the silver bullet, but I fear the likelihood of that is low.

Another idea is to have a chrome extension that updates a visual counter of unhandled/pending counter in the main chrome (maybe they appear only after X ms). This would be similar to how we realize errors are happening in our apps. This of course is only useful for browsers, and browser users using the extensions.

**Comment by stefanpenner** [comment]
for reference for anyone who hasn't seen it, this is our current progress with related browser tooling: https://github.com/tildeio/ember-extension/pull/76 

It's a start, lots more todo. If someone is interested I would love some extra bandwidth to work on having it a standalone extension (or maybe our devtools friends want to support this out-of-the-box)

**Comment by medikoo** [comment]
> A lint tool is necessary to expose such mistakes

Not true. I use lint tools heavily, but never felt a need to use lint for that and never did.

> Limited reporting when doing async error handling, e.g.

This code doesn't make much sense, and in its best interpretation behaves properly: it doesn't throw second error as fist (in order of processing) took precedence.
Same when you process collection of promises and few of them become rejected, most (probably all) libraries expose only error of first promise that was rejected, and that's totally fine, it's expected and not perceived as an issue.

**Comment by jakearchibald** [comment]
> Probably you proxy some unhandled rejections over to the console, like you do XHRs

Yes, we do this so the developer doesn't have to switch through every tab of devtools when something isn't working right. We hope to give them something in the console that gives them a clue, like we do with failed XHRs. We want to do the same with promises.

Also, this is what developers running v8 via the cmd line would see

> but you need to remove them from the console once they become handled

…which we can't do on the cmd line.

We don't want the console to become really noisy, but ideally we want to tell the developer about unhandled exceptions, like we do with sync code.

**Comment by domenic** [comment]
Yes, the command-line debugging experience is not as obvious, because it's hard to represent a live list on the command line. But for web inspector, it's a no brainer.

**Comment by stefanpenner** [comment]
@jakearchibald the catch is, this isn't sync code. Likely some timeout tolerance to decrease the likely good. (GC hook would make it 100% fine, as long as your code didn't accidentally retain the promise forever, which is totally possible)

in light of that we still likely need a combo of a few strategies to give good solution here.

**Comment by jakearchibald** [comment]
I'm not arguing against the big graphical representation of promise state, we need that and something like that will go into devtools eventually.

However, we need something for cmd-line, and if we come up with something decent, that'll also go into devtools console.

(even in graphical devtools, I don't think removing things from the console is a great experience)

**Comment by jakearchibald** [comment]
Maybe the GC option is worth a shot. Although I'm worried about inconsistencies like some errors not being logged for difficult to understand reasons (cannot be GC'd due to still-in-scope or something like eval() in scope), or errors being logged out of order.

**Comment by domenic** [comment]
I know I've said this before, but it seems like it's time to say it again. I maintain strongly that the fundamental primitive here is a live list of unhandled rejections. It is fundamentally undecidable for a computer to know what that list "should" contain: should it be empty always; should it be allowed to contain things for only one tick; should it be allowed to contain things for ten seconds; should it be allowed to contain rejections specifically generated by certain proccesses... The _only_ coherent strategy is to surface to the user the state of their system. Making a choice like "log immediately" or "log after one tick" is making an often-incorrect decision about how long things are allowed to remain in the live list.

Let us try an imprecise analogy. The list of outstanding network requests can contain connections that are kept open indefinitely, e.g. for long-polling purposes. The dev tools presents this as a live list of network requests, some of which get moved into "success" state, some of which get moved into "error" state, and some of which don't ever get moved out of the pending state, because they are used for long polling. Sometimes things stay in a pending state for a long time. But it would be crazy for the dev tools to decide "oh, this connection's been open long enough; I'll just log it as an error state and move on." Only devs can do that, by manually attaching timeouts (akin to devs manually attaching `.catch(console.error)`).

**Comment by jakearchibald** [comment]
That analogy is too imprecise. The browser logs failed network connections to the console when they fail. This gives the user a hint about the error without them going to the main network tab. I'm looking for something like that for promises that'll work when they're used in command line contexts such as node.js.

The live list is a great idea, I don't think anyone is arguing against it. However, we also need something for command-line cases, and that solution will also be useful in the devtools console.

**Comment by domenic** [comment]
I don't think a live list can be represented in a non-editable textual medium such as a traditional console log. As such, any solution that is console-only is simply insufficient for representing the potential error handling of a dynamic async system. Anything you do will be misleading.

**Comment by spion** [comment]
@medikoo a lint tool is necessary to expose such mistakes. You're claiming you don't make such mistakes - thats great! But others apparently continue to make them. Now we can argue whether the problem is education or not, but that wont change the fact that **when** a mistake is made, the user will not be warned of that mistake and instead will be greeted with the sound of silence!

Also, the second example does make sense. Imagine that both rejections are actually some asynchronous operations, but the second one attaches the handler. Lets take @stefanpenner 's example:

```
var facebook = facebookApi()
var twitter = twitterApi()

App.done(function(app){
  facebook.catch(function(error){
    app.reportFailure(error);
  }).done();

  twitter.catch(function(error){
    app.reportFailure(error);
  }).done();
});
```

Here if all the promises (App, facebook and twitter) fail, only App's failure will be reported.

**Comment by medikoo** [comment]
@spion Misuse of `done` or `then` doesn't fall into kind of mistakes that you solve with lint tools. Do you happen to  use `[].map` instead of `[].forEach` by mistake? If you do it, it's only because you don't understand the API, and no lint tool would solve that.

In your example you misuse both `catch` and `done`, it should be written that way:

``` javascript

App.done(function(app){
  facebook.done(null, function(error){
    app.reportFailure(error);
  });

  twitter.done(null, function(error){
    app.reportFailure(error);
  });
});
```

**Comment by stefanpenner** [comment]
@medikoo you are incorrect.

**Comment by spion** [comment]
@medikoo if I write a chain of `.map`s, then remove the last one, then yeah it might happen that I forget to replace the last one with `.forEach`. 

The solution isn't the bad part - the bad part is that your libary can't warn your users that they're doing it wrong when they do it wrong.

**Comment by medikoo** [comment]
@spion when you know `done` and treat it as first choice it's very unlikely to turn into corner of muted unhandled exception by mistake. Same as it's unlikely to leave `map` instead of `forEach`. I don't think I can remember such case on my side.

Anyway, even if, it totally can't be argument against `done`. It's rather argument that aside of `done`, it would be nice to have some aid in developer tools, which in such case is not a _must have_ to work reliably with promises.

**Comment by spion** [comment]
You touched on another problem there in the example - `done(null, handler);` - maybe we also need .catchdone ? :)

**Comment by medikoo** [comment]
@spion I would rather remove `catch` from spec, it's rarely used (in implementations that have `done`) and it's just sugar over `then`, brings not much value.

**Comment by spion** [comment]
Its not rarely used. You often want a function that catches certain known error conditions.

Use case: A database query execution tool has a function that executes the query and prepares a result report. It also prepares a (different) result report when the query is invalid

``` js

// not necessary with bluebird
function only(type, handler) {
  return function(e) {
    if (e instanceof type) return handler(e);
    else throw e;
  }
}

function getQueryReport(q) {
  var timing = Date.now();
  return db.query(q).then(function(results) {
    return new ResultsReport(results, timing);
  }).catch(only(QueryError, function (e) {
    // this error was expected, make sure we show it to the user
    return new ErrorReport(e, timing); 
  }));
  // all other errors are automatically propagated
}
```

But we digress.

**Comment by ForbesLindesay** [comment]
I always write `.then(null, handler)` rather than `.catch(handler)` but that's just personal preference.

Node could use a nice graphical debugger with step through and live dev tools that was good enough for everyone to use it all the time.  That still doesn't help much for surfacing errors in production systems though, which done does?

**Comment by medikoo** [comment]
On my side I'd rather make use of `then` second argument and handle errors (that are to be handled) directly after they occur.  It's unlikely that same kind of error handling will repeat across the chain, and it's much more bulletproof approach:

``` javascript
return db.query(q).then(function(results) {
  // Let eventual report error propagate freely
  return new ResultsReport(results, timing);
}, ifErr(AcceptableQueryError, queryErrorHandler));
```

If additionally if I would like to handle some specific _ResultsReport_ error, and it's sync operation I follow standard sync way to do that:

``` javascript
return db.query(q).then(function(results) {
  try {
    return new ResultsReport(results, timing);
  } catch (e) {
    // Report error
    return ifErr(AcceptableReportError, reportErrorHandler)(e);
  }
}, ifErr(AcceptableQueryError, queryErrorHandler));
```

**Comment by juandopazo** [comment]
One thing that `done` addresses very well is normalization to a callback API. Let´s say we have an API that uses promises internally. In order to expose the result of calling that API in terms of callbacks, we need to jump through some bureaucratic hoops:

``` js
function someCallbackAPI(foo, callback) {
  return doSomething(foo)
    .then(somethingElse)
    .then(moreWork)
    .then(function (result) {
      setImmediate(callback.bind(null, null, result));
    }, function (error) {
      setImmediate(callback.bind(null, error));
    });
}

function sameUsingDone(foo, callback) {
  return doSomething(foo)
    .then(somethingElse)
    .then(moreWork)
    .done(callback.bind(null, null), callback);
}
```

I don't think this use case is going away any time soon.

**Comment by joliss** [comment]
~~I don't know whether `.done` is the right solution. However,~~ I'd like to point out something to everyone who's thinking about this stuff:

Handling exceptions in asynchronous code properly is not only a usability issue. Rather, I believe that it's important from a **security** point of view. Allow me to elaborate.

Common wisdom seems to be that when there is an uncaught exception in a Node server app, keeping the server running will cause issues and leak resources, so you should simply try to gracefully shut down the server process, and start a new one. This means that if an attacker can figure out a way to trigger an uncaught exception in your app, they might be able to force you do expend a significant amount of CPU time by sending a specially crafted request. As a result, any remotely triggerable uncaught exception creates a denial of service vulnerability, because an attacker can leverage a low amount of network bandwidth into a high amount of CPU usage. A past example of this CPU-to-network leverage issue would be the slew of vulnerabilities discovered in 2011, caused by dictionaries implemented without collision-resistant hash functions; see the [original slides](http://events.ccc.de/congress/2011/Fahrplan/attachments/2007_28C3_Effective_DoS_on_web_application_platforms.pdf) on this, especially pp. 50 ff.

In a Rails app, having an uncaught exception in your simply results in a 500 page. In a Node app, it would seem to cause a vulnerability under typical conditions. This is disastrous for security, because it turns ordinary bugs into vulnerabilities.

Even worse, note that libraries tend to accumulate "dead" attack surface. For instance, for the longest time Rails was parsing YAML, even though almost nobody was using it. YAML support was what I call "dead" attack surface, because clearly it wasn't used anymore. I presume that it was kept to avoid breaking compatibility -- perhaps reasonably so. Of course YAML parsing turned out to be [ludicrously insecure](https://groups.google.com/d/msg/rubyonrails-security/KtmwSbEpzrU/NzjxkM7HLjAJ) in the end, and that's how we found out about it. But think about how many other such arcane code paths are exposed by library code in an average application. Most shouldn't contain egregious vulnerabilities of the YAML kind, but it would be very strange indeed if none had conditions under which they'd throw exceptions. Note also that it's very hard to guard against this by auditing.

To summarize, I suspect that the average Node server application contains several undiscovered DoS vulnerabilities owing to this problem. I'm not sure how to guard against this either.

My hope would be that in the future, we can use promises to evolve error handling in such a way that in most or all places, code will be able to safely throw unexpected exceptions without causing a vulnerability.

At the moment however, it seems to me that even with promises, there are still places where your code simply mustn't throw an error, and if it does, you are in a tricky spot, potentially leading to vulnerabilities as described above.

I don't have a solution, but I just wanted to point out with this comment that if we manage to solve the problem of uncaught exceptions with promises, it would not only be good for developer usability. Rather, it might also be an unexpected boon for the security of Node applications in the long run.

**tl;dr:** Node apps seem to suffer from a common class of DoS vulnerability. Promises don't avert this at the moment, but they might in the future, if we can figure out how.

_Update:_ Now regarding the specific `.done` proposal, it seems to me that it's advocating exactly the problem I described above: [Throwing unhandled exceptions into the event loop.](https://github.com/kriskowal/q/wiki/API-Reference#promisedoneonfulfilled-onrejected-onprogress) So no, `.done` is definitely not the right solution.

_Thanks to @stefanpenner for proof-reading a draft of this._

**Comment by wycats** [comment]
@joliss my thinking about what will end up being idiomatic is this:

``` js
spawn(function*() {
  let currentUser = yield User.find(currentUser);
  let post = yield Post.find(params.id, { user: currentUser });
  // more code leading up to generation of JSON, all of which uses yield
});
```

Because generators natively support proper synchronous exception handling, the class of problem that leads to swallowed errors will generally not occur when using promises together with generators. As a result, I think it would be best if we held off on making too many assumptions about how the error-swallowing behavior will affect idiomatic usage of promises, and stick with the simple one-API (`.then`).

TL;DR I think generators and `yield`ing to promises will eliminate this class of problem, so we should hurry up that future and not lard up the current API with solutions to scenarios that will go away when generators become idiomatic.

**Comment by joliss** [comment]
@wycats's elaboration from IM, in case other people are confused about how the spawn + generator thing works:

wycats: The way spawn works is
wycats: yield somePromise
wycats: Will pause the generator until the promise is resolved or rejected
wycats: If resolved, it yields the value
wycats: If rejected, it throws in the current scope
wycats: So there is no .then chaining
wycats: Ever
wycats: .then only happens internally in spawn

**Comment by spion** [comment]
@wycats unfortunately, synchronous catch in JavaScript doesn't support any kind of error filtering. Afaik, catch-all is an antipattern in all languages - one should only catch expected errors. Error bubbling makes this a real issue - at the topmost levels all kinds of errors may be aggregated and we definitely have no idea what they may all be.

Promise libraries like Bluebird alleviate this by allowing

`.catch(ErrorType | predicateFn, handler);`

But its also equally easy to write a combinator

`let ifErr = (predicate, handler) => e => { if (predicate(e)) return handler(e); throw e; }`

and use it with other libraries

`.catch(ifErr(predicateFn, handler))`

Since JS catch wont have pattern matching until ES7, we will have to resort to

`let onlyErr = (filter, e) => { if (!filter(e)) throw e; }`

and

``` js
try {
  yield fn1();
  yield fn2();
} catch (e) { onlyErr(predicate, e);
  only handle errors that satisfy the predicate
}
```

The difference is obvious - its much easier to evolve libraries. If other promise libraries find Bluebird's error filtering style useful, they can adopt it. Its equally easy for some of these libraries to even forbid `catch` without predicate (that might be a good idea for an experiment)

In contrast, it would take a huge amount of time (until ES7) for JS to incorporate error filtering, and there is unfortunately not  much opportunity for experimentation there.

I like generators, but because of the above its likely that I'll wait for ES7 before reconsidering them again, and even then I'll probably prefer the flexibility of functional style - especially after the arrival of arrows in ES6

**Comment by ForbesLindesay** [comment]
It is normal in virtually all websites to handle virtually all un-caught exceptions by sending a 500 page to the user and not re-starting the app.

**Comment by medikoo** [comment]
Recently on a local meetup I made the presentation about asynchronous interfaces in JavaScript, it covered also introduction to promises with all surrounding controversies.

One of the ideas was to explain promises starting with `done`, make it a center point of implementation, and on that basis explain `then`. It felt right logically, and in turn produced simple and very comprehensible implementation, see [medikoo/plain-promise](https://github.com/medikoo/plain-promise#plain-promise)

**Comment by kriskowal** [comment]
@medikoo I have done this refactoring in Q for the experimental branch https://github.com/kriskowal/q/blob/v2/q.js#L748-L874

**Comment by cscott** [comment]
The `prfun` library implements `Promise#done` on top of standard ES6 promises (using the `es6-shim` implementation if necessary).  It is definitely useful in the short term, although I agree with the comments above that it doesn't need to be written into the ES6 spec since it's not clear that it's the correct long-term answer.

**Comment by briancavalier** [comment]
@medikoo @kriskowal I'll pile on. when.js 3.x does basically the same thing.  It implements both `then` and `done` on top of a function that is much more like `done` than it is like `then`.

**Comment by panuhorsmalahti** [comment]
One problem with "log to console"/"log to live panel of unhandled promises" instead of .done is that if you're using .done you can catch that error and restart the Node.js process (or just let the process die as is the default). However, restarting a Node.js process from a single "unhandled promise logged to the console" is probably not the default. In addition, from the discussion above it seems that there will be false positives, and that probably leads to most people not enabling the "restart" functionality in production or otherwise.

The result of this, I fear, is that there will be production and development Node.js applications that contain errors that are not noticed opening security issues.

**Comment by benjamingr** [comment]
@panuhorsmalahti actually I performed research on that today for Bluebird (which offers both options and also lets you set what happens when an `possiblyUnhandledRejection` occurs) - most people (around 70%) logged an unhandled rejection (and didn't crash), a bunch (around 20%) _did_ crash the process.

In any way the errors will not be "suppressed".

**Comment by xtianus79** [comment]
Anymore debate on this .done() topic?  Very interesting read.  Once question I have which may be stupidly obviously but is the use of .done to replace or finish a .then() chain?  .then().then().then().done(); <<< ??? 

Or can you chain them?  Perhaps I am just thinking of this wrong but it is mentioned many times but not well understood I guess for me of ... "know where and when to use .done()?"  

Just wanted to update: 

I found this site here which explained it very well... and I get the point of what he( @ForbesLindesay ) is trying to say... 

In other words, .done() is not using a promise which requires a value but rather a returning function based on it's own merits???  Is done semantically the correct phrase here?  it implies a few things that I venture to be ambiguous. 
1.  The promise chain/transaction is finished... is that true?  Can you chain done()'s?  Can you make a done(). followed by a then()? would you ever want to do that?  
2.  It does not imply that you can't use the value/result of the preceding promise.  

(( I would call them loaded promises and empty promises(haha) no... seriously unloaded promises.  ))

In reality couldn't I then have a .thenL and .thenUl for my purposes?  I get the .done this supposed to fire off an un-handled error into the event loop but in chaining purposes what if you want that in general.  Again, am I supposed to chain .done()'s? that is very confusing and not very semantic.

if the pure purpose is to use an unnecessary resulted promise to then have an error come to the event loop this naming convention is odd.  If it is to do this purely at the end of a promise chain which is then to use an unnecessary resulted promise that reports to the event loop --- well isn't that a catch?   

One last point, with the .thenL() and .thenUl() framework the chaining becomes obvious but also a .thenDone() could finalize and finish the chain operation that then has any other special functionality besides being at the end and being done and being able to report an error to the event loop?  

This could prevent promise nesting hell.  Your thoughts?   

reference: 
https://www.promisejs.org/

> Put simply, .then is to .done as .map is to .forEach. To put that another way, use .then whenever you're going to do something with the result (even if that's just waiting for it to finish) and use .done whenever you aren't planning on doing anything with the result.

**Comment by xtianus79** [comment]
Wow this kind of throws cold water on that whole thing no or this just in relation to bluebird?  

https://github.com/petkaantonov/bluebird/issues/471

**Comment by medikoo** [comment]
@xtianus79 `done` simply is just plain way to access resolved value. So if you do not need a promise returned by `then` (in other words: you don't need to extend then chain), then it's cleaner to use `done`. It shouldn't be understood anything more than that.

This issue also promotes `done` as a solution for error swallowing, but I agree now, that introduction of `done` doesn't solve this issue well. Firstly there are cases where superfluous `done()` calls still will be required from a programmer (e.g. `writeFile(path, content).done()`), and secondly programmer shouldn't be punished for using `then` mistakenly instead of `done` (same as he/she is not punished when using `map` over `forEach` when there's no need for)

**Comment by ForbesLindesay** [comment]
To answer your main question: no, you cannot chain `.done`s.

You suggest that the normal behaviour you want is for errors to be re-thrown/reported.  I dispute that claim somewhat.  The normal behaviour you want is for all your errors to clime the stack of functions that called the code that threw an error.  I don't want an error reported in readFile if I later handle it in `readSettingsFile` by simply returning the default settings.  The time I **do** want error to be re-thown/reported is at the top of the chain.  For example, someone clicks a "vote" button in my app:

``` js
voteButton.addEventListener('click', function () {
  request('post', '/api/vote').then(function () { // here I'm transforming the result of the first request, by making a second request
    return request('get', '/api/updated-data');
  }).done(function (data) { // here, I'm done adding handlers.  The last thing I'm doing is updating the UI.
    updateUI(data);
  });
});
```
